src/main/scala/rocket/Rocket.scala

```diff
@@ -213,10 +213,10 @@ class Rocket(implicit p: Parameters) extends CoreModule()(p)
   val id_amo_rl = id_inst(0)(25)
   val id_fence_next = id_ctrl.fence || id_ctrl.amo && id_amo_rl
   val id_mem_busy = !io.dmem.ordered || io.dmem.req.valid
+  when (!id_mem_busy) { id_reg_fence := false }
   val id_rocc_busy = Bool(usingRoCC) &&
     (io.rocc.busy || ex_reg_valid && ex_ctrl.rocc ||
      mem_reg_valid && mem_ctrl.rocc || wb_reg_valid && wb_ctrl.rocc)
-  id_reg_fence := id_fence_next || id_reg_fence && id_mem_busy
   val id_do_fence = id_rocc_busy && id_ctrl.fence ||
     id_mem_busy && (id_ctrl.amo && id_amo_aq || id_ctrl.fence_i || id_reg_fence && (id_ctrl.mem || id_ctrl.rocc))
 
```

Comment:
I didn't write the question properly - of course, if illegal instruction is executed and exception is taken, then stalls don't matter.

What I am trying to figure out is when fence is supposed to happen according to `aq` and `rl` modifiers.

Here is a waveform before this PR:
![id_ctrl_amo and id_ctrl_mem](https://cloud.githubusercontent.com/assets/16438432/24675433/f8beb62a-1933-11e7-919f-9cac4d8942dd.png)

At C1 there is no illegal instruction exception, `id_ctrl.legal == 0`, `id_ctrl.mem == 0`, however, `id_ctrl.amo == 1`, `id_amo_aq == 1`, `id_amo_rl == 1` and `id_fence_next == 1`. And if `id_mem_busy` is asserted, then fence should happen according to the logic above.